### ISSUE TITLE
[BUG] make_reduction using X, y and window_length=1 DOESN'T consider X[today] to predict y[today]

### ISSUE THREAD

**Issue by corradomio** [sktime/sktime#4511]
**Describe the bug**
'make_reduction' uses 'scikit-learn' compatible models to create forecasters.

We suppose to have:

    - input/exogenous variables X (n x 2) and target y (n x 1) 
    - window-length = 1
    - model used: LinearRegression

To 'predict' the TODAY's target value 'y' (y[0]), we have to use:

    - X[-1]: input values of YESTERDAY
    - y[-1]: target value of YESTERDAY
    - X[ 0]: input values of TODAY

This means than the transformed matrix used to train the LinearRegression model must have

    2+1+2 = 5

five columns.

This is 'consistent' with a window-length of 0: indeed, in this case the forecaster 's behavior 
is exactly the same of the regressor.

**To Reproduce**

```python
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sktime.forecasting.compose import make_reduction

x = np.arange(100)
y = np.arange(-50, 50)
z = x + y

df = pd.DataFrame({'x': x, 'y': y, 'z': z})
tr = df[:80]
te = df[80:]

f = make_reduction(LinearRegression(), window_length=1, strategy='recursive')
f.fit(X=tr[['x', 'y']], y=tr['z'])

```
to add a breakpoint in [PYTHON-LIBRARIES]\sktime\forecasting\compose\_reduce.py line 565, 
immediately after the following line:
 
      yt, Xt = self._transform(y, X)

      # Make sure yt is 1d array to avoid DataConversion warning from scikit-learn.
     if self.transformers_ is not None:    <<< [BREAKPOINT]

The content of Xt (3 columns) is

  z     x     y
----|----|----
2.0, 1.0, 1.0
4.0, 2.0, 2.0
6.0, 3.0, 3.0
8.0, 4.0, 4.0
 ...

**Expected behavior**
To predict y[0] (TODAY) we have to use X[-1], y[-1] (YESTERDAY) AND X[0] (TODAY)

z[-1], x[-1], y[-1], x[0], y[0]
------------------------------
2.0,   1.0,  1.0,   2.0,   2.0
...

**Versions**
<details>
System:
    python: 3.9.16 (main, Mar  8 2023, 10:39:24) [MSC v.1916 64 bit (AMD64)]
executable: D:\Python\Anaconda3-2022.05\envs\torchv2\python.exe
   machine: Windows-10-10.0.19045-SP0

Python dependencies:
          pip: 23.0.1
       sktime: 0.17.0
      sklearn: 1.2.2
        numpy: 1.23.5
        scipy: 1.10.1
       pandas: 1.5.3
   matplotlib: 3.7.1
       joblib: 1.2.0
  statsmodels: 0.13.5
        numba: 0.56.4
     pmdarima: None
      tsfresh: None
   tensorflow: None
tensorflow_probability: None
</details>

**Comment by fkiraly** [comment]
@corradomio, I think the implementation is correct, albeit the use of exogeneous data is nowhere properly documented - we should probably do that.

The algorithm that I think should be used is below, as in the docstring of the `RecursiveReductionForecaster` that is still experimental at the moment:

```

    In `fit`, given endogeneous time series `y` and possibly exogeneous `X`:
        fits `estimator` to feature-label pairs as defined as follows.

        features = `y(t)`, `y(t-1)`, ..., `y(t-window_size)`, if provided: `X(t+1)`
        labels = `y(t+1)`
        ranging over all `t` where the above have been observed (are in the index)

    In `predict`, given possibly exogeneous `X`, at cutoff time `c`,
        applies fitted estimators' predict to
        feature = `y(c)`, `y(c-1)`, ..., `y(c-window_size)`, if provided: `X(c+1)`
        to obtain a prediction for `y(c+1)`.
        If a given `y(t)` has not been observed, it is replaced by a prediction
        obtained in the same way - done repeatedly until all predictions are obtained.
        Out-of-sample, this results in the "recursive" behaviour, where predictions
        at time points c+1, c+2, etc, are obtained iteratively.
        In-sample, predictions are obtained in a single step, with potential
        missing values obtained via the `impute` strategy chosen.
```

Note that here, `X` is not getting lagged, so in your above example of `window_size`=0, this algorithm would predict `y[0]` (today) from `X[0]` (today) and `y[-1]` (yesterday).

The reason for this is, if you want `X` to be lagged, you can pipeline the reducer with a `Lag` transformer (on the eogeneous data, by `ForecastingPipeline` or the `**` dunder).

Whereas, if the reducer would already lag the `X`, there would be no easy way to ensure the prediction does *not* use `X[-1]`.

Please let me know if you disagree with the "should" or the "is".
Also FYI @danbartl.

**Comment by corradomio** [comment]
Hi @fkiraly, there is a 'philosophical problem', I suppose ;-)
Obviously, the implementation must be consistent with the Time-series Analysis concepts ;-)

The time series model:  y[0] = a*X[0] + b1*y[-1] + b2*y[-2]+...  is the 'classic' model.

However it is ""reasonable"" to consider the model:

y[0] = a0*X[0] + a1*X[-1] + b1*y[-1] + a2*X[-2] + b2*y[-2] + ....

(this is exactly the model I am implementing for a paper)

At this point, there are two ideas:

1. "windows_size = 0" means to use ONLY X[0] (if present) and y[-1], BUT in this case it must be possible to specify "windows_size=-1"
2. "windows_size = 0" means to use ONLY X[0]. 
 
You can consider the ARIMA definition (https://en.wikipedia.org/wiki/Autoregressive_integrated_moving_average):
the parameters 'p' and 'q' must be >= 1 because the sums start with i=1. If we consider 'window_size' as 'p' or 'q', 'windows_size=0' means don't use the sums ;-).

The other problem is: WHY to write (ws: window_size)

a) y[c+1] = const +  a_0*y[c] + ....  a_ws*y[c - ws]

and NOT

b) y[c] = const + a_1*y[c-1] + ... a_ws*y[c - ws]

('const' can be 0)

???

From a mathematical point of view, the two equations are exactly the same, 
BUT they deffer on the possible values for 'ws'.
To ""avoid" the sum 'ws' must be

a),  '-1', 
b) '0', MORE reasonable value.

**Comment by corradomio** [comment]
Extra note: the current implementation of 'RecursiveTabluarRegressionForecaster' (RTRF) implements exactly the model 

y[0] = a_0 X[0] + a_1 X[-1] + b_1 y[-1] + a_2 X[-2] + b_2 y[-2] + ....

except for the missing X[0] ;-)
I know it because I am implementing a little more complex variant and I am using the RTRF as 'reference'.

**Comment by fkiraly** [comment]
thanks for clarifying - I think there is a clear need to document what precisely the estimator does, at a minimum.

There also seem to be two expectations, so instead of taking a philosophical stance for one or the other, that could simply be a paramater?

**Comment by corradomio** [comment]
In programming, the parameters are the solution for 'all philosophical problems' ;-)

**Comment by fkiraly** [comment]
so, what I'll do is:

* clarify the docstrings in the reducer(s)
* explain how to lag `X` with the current reducer, and add a cautionary note that it will not to under default parameters

would that be a good solution, @corradomio?

**Comment by corradomio** [comment]
@fkiraly, I am sorry, BUT it is possible that I described the problem in the wrong way.
There are 2 possible, and perfectly VALID, TS models:

y[0] = TSM1( X[0], y[-1], X[-1], y[-2], X[-2], ...  )

AND

y[0] = TSM2( X[0], y[-1], y[-2], ...)

I am sure at 99% ( ;-) ) that the model:

y[0] = TSM3( y[-1], X[-1], y[-2], X[-2], ... )   (it DOESN'T consider X[0])

is WRONG.
This is confirmed by equation 13.2, pag. 212 of Applied Time Series Analysis (Terence C. Miller, Academic Press, 2019).
(it is not difficult to find the PDF ;-) )

---

Now, the class RecursiveTabularRegressionForecaster (result of the last Python code posted) implements the model TSM3 instead than TSM1.

Then, I think, there is a bug in **RecursiveTabularRegressionForecaster**:

1. it DOESN'T implement TSM2, because it implements TSM3, OR
2. it implements TSM1 in WRONG way,

---

If you are sure that TSM3 is valid, at minimum, in the documentation, it is necessary to specify a 'reference' to a paper or Wiki page that describe the model (to use software in research papers, it is mandatory to specify the references of the models used).

The "extra" solution is to use some flag to select one between TSM1, TSM2 and TSM3.

**Comment by fkiraly** [comment]
> at minimum, in the documentation, it is necessary to specify a 'reference' to a paper or Wiki page that describe the model (to use software in research papers, it is mandatory to specify the references of the models used).

Yes, I fully agree to this one!

I think the problem is primarily one of documentation, and one of "which model is common" second.

We know that documentation is not great in places, but `sktime` is open source and volunteer driven.

The entire reason why I started writing `RecursiveTabularRegressionForecaster` is to increase clarity of code structure and accuracy of documentation.

If you would be willing to help with cleaning up the module, you would be making a huge contribution!!
(that can be anything from docstrings to code implementation, speed up, etc)

From a philosophy perspective, `sktime` understands itself as a library, i.e., any algorithm can be added as long as it is clearly and properly documented, and has either a full formal description in place, or by reference. So, if you were to contribute sth, or want to make improvements, it would be much appreciated.

**Comment by fkiraly** [comment]
Next thing, may I clarify your understanding? I'm just trying to repeat precisely what you say for confirmation, can you please let me know whether you think I understood correctly?

You think:
* `RecursiveTabularRegressionForecaster` implements TSM3
* `RecursiveReductionForecaster` implements TSM2
* `RecursiveTabularRegressionForecaster` should implement TSM1 because users will expect it implements TSM1 rather than TSME3, and/or because TSM1 is more common

Also, can I confirm whether you agree:
* `RecursiveReductionForecaster` pipelined with `Lag` or `WindowSummarizer` can be used to obtain TSM1 or TSM3 (depending on lag parameters chosen)
* users should be able to choose the lags for `X` independently from the lags for `y` because that is common in literature

**Comment by corradomio** [comment]
1. exactly
2. I have to investigate
3. TSM1 is the TS model I found in several books. I don't know if it is "more common" or not
4. I think it is a good idea
5. to use different lags (just an integer value) for X and y is just the FIRST solution.

I am working on a model where it is possible to specify lags for X (input) and y (target) as a list of integers:

```
lag_for_X=[0,1,3,7,14,21,30]
lag_for_y=[1,3,5]

--> y[0] = f(X[0], X[-1], X[-3], X[-7], ..., y[-1], y[-3], y[-5])
```

In this way, if you write, for example:

```
lag=3     --> lag_for_X=[0,1,2,3], lag_for_y=[1,2,3]
lag=(3,2) --> lag_for_X=[0,1,2,3], lag_for_y=[1,2]
```

Unfortunately, in this moment, the solution is not 'sktime-compliat'.

But I am sure that you known a lot more than me, Dr. Király ;-)

**Comment by fkiraly** [comment]
> The idea to replace my implementation with an 'extended version' 

I think that's what I've tried in `RecursiveReductionForecaster`, but the runtime is slow, for reasons we haven't fully tracked down.

The prototype is in the release (just in a private module) so you could play around with it.
https://github.com/sktime/sktime/issues/3224
If you could track down what slows it down so much, that would be hugely helpful - then you might be able to have your cake and eat it?

Of course, if you want to contribute your solution to `sktime` in a compliant interface, we're also happy to help with that.

**Comment by round80** [comment]
is there currently a way to only create lag for y but not x?

**Comment by gygabyte017** [comment]
> `RecursiveReductionForecaster` pipelined with Lag or WindowSummarizer can be used to obtain TSM1 or TSM3 (depending on lag parameters chosen)

@fkiraly can you provide an example to obtain a TSM3? Tried like

```
forecaster = TransformedTargetForecaster(
    [
        ("lag", Lag(lags=1, keep_column_names=True, index_out="original")),
        ("forecast", make_reduction(LinearRegression(), window_length=1, pooling="global", strategy="recursive"),
        ),
    ]
)
```

But I'm getting NaN issues (which are normally handled by the reduction).
The goal is to have a simple model like:

```
Y[t] = Y[t-1] + X[t] + X[t-1]
            ^ (*)         ^^^ (**)
```

(*) To have this I'm using the reducer with window=1 and pooling=global
(**) I would like to have this trying with Lag but to now no luck

Thank you
